### PR TITLE
[Mqtt Agent] MQTT Command Loop Complete Coverage

### PR DESCRIPTION
**Pull request by gshvang**
PR for 100% coverage of MQTT_Command Loop function.

### CODE REVIEW

File: test/unit-test/mqtt_agent_utest.c [hunk 1]
```diff
@@ -258,6 +269,45 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
     return MQTTSuccess;
 }
 
+/**
+ * @brief A stub for MQTT_ProcessLoop function to be used to test multiple calls to MQTT_ProcessLoop.
+ */
+MQTTStatus_t MQTT_ProcessLoop_stub_multiple( MQTTContext_t * pContext,
+                                             uint32_t timeoutMs )
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:276]: This needs an additional parameter to match CMock's stub function signature
```suggestion
                                             uint32_t timeoutMs,
                                             int NumCalls )
```

File: test/unit-test/mqtt_agent_utest.c [hunk 2]
```diff
@@ -258,6 +269,45 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
...
+        status = MQTTSuccess;
+    }
+    else if( processLoopCounter == 1 )
+    {
+        pMqttAgentContext->packetReceivedInLoop = true;
+        status = MQTTRecvFailed;
+    }
+    else
+    {
+        pMqttAgentContext->packetReceivedInLoop = false;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:302]: When will this `else` ever execute? If the loop counter is 1 then `MQTTRecvFailed` is returned and the process loop won't be called again

File: test/unit-test/mqtt_agent_utest.c [hunk 3]
```diff
@@ -1104,22 +1154,25 @@ void test_MQTTAgent_CommandLoop_with_empty_command_queue( void )
...
-    pReturnFlags.endLoop = true;
+    pReturnFlags.endLoop = false;
 
     MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
     MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &pReturnFlags );
     MQTT_ProcessLoop_IgnoreAndReturn( MQTTSuccess );
 
+    MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTBadParameter );
+    MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &pReturnFlags );
+    MQTT_ProcessLoop_IgnoreAndReturn( MQTTBadParameter );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1171]: With this, lines 1165-1167 become unused. Also, we should probably be in the habit of using expect rather than ignore for failure cases or when we expect to call it exactly once/a few times, both of which apply here

File: test/unit-test/mqtt_agent_utest.c [hunk 4]
```diff
@@ -1305,6 +1354,19 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
     /* Ensure that callback is invoked. */
     TEST_ASSERT_EQUAL( 2, commandCompleteCallbackCount );
 
+    /* Invoking mqttEventCallback with MQTT_PACKET_TYPE_CONNECT packet type to cover default case. */
+    packetType = MQTT_PACKET_TYPE_CONNECT;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1358]: This is literally not possible. If there's a path in the switch that can't be covered otherwise then that should be a code change.

File: test/unit-test/mqtt_agent_utest.c [hunk 5]
```diff
@@ -1355,7 +1417,7 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
 
     TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
     /* Ensure that callback is invoked. */
-    TEST_ASSERT_EQUAL( 6, commandCompleteCallbackCount );
+    TEST_ASSERT_EQUAL( 7, commandCompleteCallbackCount );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1420]: We should write our tests such that rearranging some or inserting one in the middle shouldn't require incrementing a variable in every assert that follows. If they're testing separate things, they should be separate tests, or this variable should be reset between API calls.

File: test/unit-test/mqtt_agent_utest.c [hunk 6]
```diff
@@ -1416,5 +1478,39 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    /* Initializing command to be sent to the commandLoop. */
+    commandToSend.commandType = PUBLISH;
+    commandToSend.pCommandCompleteCallback = stubCompletionCallback;
+    commandToSend.pCmdContext = NULL;
+    commandToSend.pArgs = NULL;
+    mqttAgentContext.agentInterface.pMsgCtx->pSentCommand = &commandToSend;
+
+    mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
+
+    TEST_ASSERT_EQUAL( MQTTRecvFailed, mqttStatus );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1513]: Why should this return `MQTTRecvFailed`? Everything else in this test implies this should be successful.

File: test/unit-test/mqtt_agent_utest.c [hunk 7]
```diff
@@ -1416,5 +1478,39 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    setupAgentContext( &mqttAgentContext );
+    mqttAgentContext.mqttContext.connectStatus = MQTTConnected;
+    pReturnFlags.addAcknowledgment = false;
+    pReturnFlags.runProcessLoop = true;
+    pReturnFlags.endLoop = true;
+
+    MQTTAgentCommand_Publish_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTTAgentCommand_Publish_ReturnThruPtr_pReturnFlags( &pReturnFlags );
+    /* Calling MQTT_ProcessLoop multiple times. */
+    MQTT_ProcessLoop_StubWithCallback( MQTT_ProcessLoop_stub_multiple );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1502]: `_StubWithCallback` is deprecated according to [CMock documentation](https://github.com/ThrowTheSwitch/CMock/blob/master/docs/CMock_Summary.md#callback). This should be `_Stub` if we haven't added any expect calls for `MQTT_ProcessLoop` and don't want CMock to keep track of the number of invocations; if we do, then it's `_AddCallback`
```suggestion
    MQTT_ProcessLoop_Stub( MQTT_ProcessLoop_stub_multiple );
```

File: test/unit-test/mqtt_agent_utest.c [hunk 8]
```diff
@@ -258,6 +269,45 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
...
+    pMqttAgentContext->packetReceivedInLoop = false;
+
+    if( processLoopCounter == 0 )
+    {
+        pMqttAgentContext->packetReceivedInLoop = true;
+        status = MQTTSuccess;
+    }
+    else if( processLoopCounter == 1 )
+    {
+        pMqttAgentContext->packetReceivedInLoop = true;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:297]: You don't need to set this flag since it's set when you call `pContext->appCallback()`. It would make more sense to just not call the callback when you want to set it to false.

**Review by muneebahmed10 — CHANGES_REQUESTED** [review]

File: test/unit-test/mqtt_agent_utest.c [hunk 9]
```diff
@@ -258,6 +270,37 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
...
+
+    pContext->appCallback( pContext, &packetInfo, &deserializedInfo );
+
+    if( processLoopCounter == 0 )
+    {
+        status = MQTTSuccess;
+    }
+    else
+    {
+        status = MQTTRecvFailed;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:296]: Comment why we're returning failure here

File: test/unit-test/mqtt_agent_utest.c [hunk 10]
```diff
@@ -258,6 +270,37 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
...
+{
+    MQTTPacketInfo_t packetInfo;
+    MQTTDeserializedInfo_t deserializedInfo;
+    MQTTAgentContext_t * pMqttAgentContext;
+    MQTTStatus_t status;
+
+    packetInfo.type = packetType;
+    deserializedInfo.packetIdentifier = packetIdentifier;
+
+    pContext->appCallback( pContext, &packetInfo, &deserializedInfo );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:288]: I thought you were only going to call this based on the loop counter?

File: test/unit-test/mqtt_agent_utest.c [hunk 11]
```diff
@@ -1104,22 +1147,23 @@ void test_MQTTAgent_CommandLoop_with_empty_command_queue( void )
...
+    mqttAgentContext.mqttContext.connectStatus = MQTTNotConnected;
     pReturnFlags.addAcknowledgment = false;
     pReturnFlags.runProcessLoop = true;
-    pReturnFlags.endLoop = true;
+    pReturnFlags.endLoop = false;
 
     MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
     MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &pReturnFlags );
-    MQTT_ProcessLoop_IgnoreAndReturn( MQTTSuccess );
+    MQTT_ProcessLoop_Stub( MQTT_ProcessLoop_stub );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1160]: What's the point of this stub if it doesn't do anything?

File: test/unit-test/mqtt_agent_utest.c [hunk 12]
```diff
@@ -1104,22 +1147,23 @@ void test_MQTTAgent_CommandLoop_with_empty_command_queue( void )
...
+
+    MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTBadParameter );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1162]: It really shouldn't be returning bad parameter when all the parameters are valid. Receive failed makes more sense

File: test/unit-test/mqtt_agent_utest.c [hunk 13]
```diff
@@ -1104,22 +1147,23 @@ void test_MQTTAgent_CommandLoop_with_empty_command_queue( void )
...
 
     mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
 
-    TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1166]: What was the reason to test for failure here? A comment would be helpful

File: test/unit-test/mqtt_agent_utest.c [hunk 14]
```diff
@@ -1173,13 +1216,12 @@ void test_MQTTAgent_CommandLoop_add_acknowledgment_success( void )
...
     MQTTStatus_t mqttStatus;
     MQTTAgentContext_t mqttAgentContext;
-    MQTTAgentCommandFuncReturns_t pReturnFlags;
     Command_t commandToSend;
 
     setupAgentContext( &mqttAgentContext );
     mqttAgentContext.mqttContext.connectStatus = MQTTConnected;
     pReturnFlags.addAcknowledgment = true;
-    pReturnFlags.runProcessLoop = true;
+    pReturnFlags.runProcessLoop = false;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1224]: If this is false do you need the ignore and return?

File: test/unit-test/mqtt_agent_utest.c [hunk 15]
```diff
@@ -258,6 +270,37 @@ MQTTStatus_t MQTT_ProcessLoop_stub( MQTTContext_t * pContext,
...
+    MQTTDeserializedInfo_t deserializedInfo;
+    MQTTAgentContext_t * pMqttAgentContext;
+    MQTTStatus_t status;
+
+    packetInfo.type = packetType;
+    deserializedInfo.packetIdentifier = packetIdentifier;
+
+    pContext->appCallback( pContext, &packetInfo, &deserializedInfo );
+
+    if( processLoopCounter == 0 )
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:290]: Do we even need this when `NumCalls` exists?

File: test/unit-test/mqtt_agent_utest.c [hunk 16]
```diff
@@ -1380,41 +1413,68 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
     TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
     /* Ensure that callback is not invoked. */
-    TEST_ASSERT_EQUAL( 9, commandCompleteCallbackCount );
+    TEST_ASSERT_EQUAL( 0, commandCompleteCallbackCount );
+}
+
+/**
+ * @brief Test MQTT_ProcessLoop failure case via calling it multiple times.
+ */
+void test_MQTT_ProcessLoop_failure( void )
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1460]: This name can be more descriptive as to what we're testing
```suggestion
void test_MQTTAgent_CommandLoop_ProcessLoop_failure( void )
```

File: test/unit-test/mqtt_agent_utest.c [hunk 17]
```diff
@@ -93,6 +93,11 @@ static uint8_t packetType;
  */
 static uint16_t packetIdentifier;
 
+/**
+ * @brief Return flags to use for test.
+ */
+static MQTTAgentCommandFuncReturns_t pReturnFlags;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:99]: Should this be memset to 0 for every test?

**Review by muneebahmed10 — COMMENTED** [review]
I believe this needs a rebase with the latest `main` branch?

File: test/unit-test/mqtt_agent_utest.c (hunk 17 above)
Comment (gshvang) [reply test/unit-test/mqtt_agent_utest.c:99]: It is not a pointer. I have changed the name.

File: test/unit-test/mqtt_agent_utest.c (hunk 10 above)
Comment (gshvang) [reply test/unit-test/mqtt_agent_utest.c]: No, its need to be called in both cases for coverage.

File: test/unit-test/mqtt_agent_utest.c [hunk 18]
```diff
@@ -1102,22 +1137,21 @@ void test_MQTTAgent_CommandLoop_with_empty_command_queue( void )
...
+    mqttAgentContext.mqttContext.connectStatus = MQTTNotConnected;
+    returnFlags.addAcknowledgment = false;
+    returnFlags.runProcessLoop = true;
+    returnFlags.endLoop = false;
 
     MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
-    MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &pReturnFlags );
-    MQTT_ProcessLoop_IgnoreAndReturn( MQTTSuccess );
+    MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &returnFlags );
+    MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTRecvFailed );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1150]: Looking at MQTTAgentCommand_ProcessLoop() it can't return anything except MQTTSuccess: https://github.com/FreeRTOS/coreMQTT-Agent/blob/560ce2cc9c60e6aa1be7d19a62a2d95eb2e00b54/source/mqtt_agent_command_functions.c#L55
We should probably be making sure the values we use in our unit tests are actually possible. If this is for coverage, we should be testing paths where it is actually possible for there to be an error. If we need to write another test case that's specifically for this situation then we can just do so.

File: test/unit-test/mqtt_agent_utest.c (hunk 17 above)
Comment (muneebahmed10) [reply test/unit-test/mqtt_agent_utest.c]: It not being a pointer doesn't affect whether we should reset it or not. If we use it for multiple tests we should probably reset it.

File: test/unit-test/mqtt_agent_utest.c (hunk 18 above)
Comment (gshvang) [reply test/unit-test/mqtt_agent_utest.c]: Added a new test case for coverage.

File: test/unit-test/mqtt_agent_utest.c [hunk 19]
```diff
@@ -1299,128 +1330,194 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    setupAgentContext( &mqttAgentContext );
+
+    mqttAgentContext.mqttContext.connectStatus = MQTTNotConnected;
+    returnFlags.addAcknowledgment = false;
+    returnFlags.runProcessLoop = false;
+    returnFlags.endLoop = false;
+
+    /* Initializing command to be sent to the commandLoop. */
+    commandToSend.commandType = PUBLISH;
+    commandToSend.pCommandCompleteCallback = NULL;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1509]: You should probably add the stub callback and check that it's invoked twice

File: test/unit-test/mqtt_agent_utest.c [hunk 20]
```diff
@@ -1299,128 +1330,194 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+ */
+void test_MQTTAgent_CommandLoop_with_multiple_commands( void )
+{
+    MQTTStatus_t mqttStatus;
+    MQTTAgentContext_t mqttAgentContext;
+    Command_t commandToSend = { 0 };
+
+    setupAgentContext( &mqttAgentContext );
+
+    mqttAgentContext.mqttContext.connectStatus = MQTTNotConnected;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1502]: This doesn't look useful for this test
```suggestion
```

**Review by muneebahmed10 — APPROVED** [review]

File: source/mqtt_agent.c [hunk 21]
```diff
@@ -657,7 +657,6 @@ static void mqttEventCallback( MQTTContext_t * pMqttContext,
                  * PINGRESP with the use of MQTT_ProcessLoop API function. */
                 LogWarn( ( "PINGRESP should not be handled by the application "
                            "callback when using MQTT_ProcessLoop.\n" ) );
-                break;
```
Comment (aggarg) [inline source/mqtt_agent.c:660]: Why is this removed?

File: test/unit-test/mqtt_agent_utest.c [hunk 22]
```diff
@@ -270,6 +278,36 @@ MQTTStatus_t MQTT_ProcessLoop_CustomStub( MQTTContext_t * pContext,
     return MQTTSuccess;
 }
 
+/**
+ * @brief A stub for MQTT_ProcessLoop function to be used to test multiple calls to MQTT_ProcessLoop.
+ */
+MQTTStatus_t MQTT_ProcessLoop_stub_multiple( MQTTContext_t * pContext,
+                                             uint32_t timeoutMs,
+                                             int NumCalls )
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:286]:
```suggestion
                                             int numCalls )
```

File: test/unit-test/mqtt_agent_utest.c [hunk 23]
```diff
@@ -1299,128 +1330,197 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    setupAgentContext( &mqttAgentContext );
+
+    mqttAgentContext.mqttContext.connectStatus = MQTTConnected;
+    returnFlags.addAcknowledgment = false;
+    returnFlags.runProcessLoop = true;
+    returnFlags.endLoop = false;
+
+    MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &returnFlags );
+    /* Calling MQTT_ProcessLoop multiple times and returning failure second time. */
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:1483]: I see that `MQTT_ProcessLoop_Stub` is only called once while the comment says that it is called multiple times. Does `MQTTAgent_CommandLoop` call it? If so, would you please update the comment. Else, please update the test.

File: test/unit-test/mqtt_agent_utest.c [hunk 24]
```diff
@@ -1299,128 +1330,197 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+
+    mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
+
+    TEST_ASSERT_EQUAL( MQTTRecvFailed, mqttStatus );
+}
+
+/**
+ * @brief Test MQTTAgent_CommandLoop behavior processing multiple commands in command queue.
+ */
+void test_MQTTAgent_CommandLoop_with_multiple_commands( void )
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:1494]: What is this test checking - 1. Second command fails to get posted to the queue. OR 2. Second command gets posted to the queue and fails while processing? Would you please update the test name and comment accordingly.

File: source/mqtt_agent.c (hunk 21 above)
Comment (gshvang) [reply source/mqtt_agent.c:660]: To cover the default case as it was not being covered otherwise by any other unit test case.

File: test/unit-test/mqtt_agent_utest.c (hunk 23 above)
Comment (gshvang) [reply test/unit-test/mqtt_agent_utest.c:1483]: Yeah MQTTAgent_CommandLoop call it. I have updated the comment.

File: test/unit-test/mqtt_agent_utest.c (hunk 24 above)
Comment (gshvang) [reply test/unit-test/mqtt_agent_utest.c:1494]: It is testing the second scenario. I have updated the comment.

File: source/mqtt_agent.c (hunk 21 above)
Comment (aggarg) [reply source/mqtt_agent.c:660]: Will this will result in contradicting messages - One, saying that `PINGRESP` is received and should not be handled and Second, saying that unknown packet type received?

File: test/unit-test/mqtt_agent_utest.c [hunk 25]
```diff
@@ -1299,128 +1330,197 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    returnFlags.addAcknowledgment = false;
+    returnFlags.runProcessLoop = true;
+    returnFlags.endLoop = false;
+
+    MQTTAgentCommand_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTTAgentCommand_ProcessLoop_ReturnThruPtr_pReturnFlags( &returnFlags );
+    /* MQTTAgent_CommandLoop calling MQTT_ProcessLoop multiple times and returning failure second time. */
+    MQTT_ProcessLoop_Stub( MQTT_ProcessLoop_stub_multiple );
+
+    mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:1486]:
```suggestion
    /* The following stub ensures that the second or later call to
     * MQTT_ProcessLoop will fail. */
    MQTT_ProcessLoop_Stub( MQTT_ProcessLoop_FailSecondAndLaterCallsStub );

    /* MQTTAgent_CommandLoop keeps calling MQTT_ProcessLoop until an error occurs.
     * In this case, the second call to MQTT_ProcessLoop will error out (as ensured
     * by the above stub) and therefore, we expect the MQTTAgent_CommandLoop
     * to fail.*/
    mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
```

File: test/unit-test/mqtt_agent_utest.c [hunk 26]
```diff
@@ -1299,128 +1330,197 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+
+    TEST_ASSERT_EQUAL( MQTTRecvFailed, mqttStatus );
+}
+
+/**
+ * @brief Test MQTTAgent_CommandLoop failure while processing multiple commands in command queue.
+ */
+void test_MQTTAgent_CommandLoop_failure_processing_multiple_commands( void )
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:1494]:
```suggestion
/**
 * @brief Test MQTTAgent_CommandLoop failure when executing second command in
 * the command queue fails.
 */
void test_MQTTAgent_CommandLoop_failure_executing_second_command( void )
```

File: test/unit-test/mqtt_agent_utest.c [hunk 27]
```diff
@@ -1299,128 +1330,197 @@ void test_MQTTAgent_CommandLoop_with_eventCallback( void )
...
+    commandToSend.pCommandCompleteCallback = stubCompletionCallback;
+    commandToSend.pCmdContext = NULL;
+    commandToSend.pArgs = NULL;
+
+    mqttAgentContext.agentInterface.pMsgCtx->pSentCommand = &commandToSend;
+
+    MQTTAgentCommand_Publish_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTTAgentCommand_Publish_ReturnThruPtr_pReturnFlags( &returnFlags );
+    MQTT_ProcessLoop_ExpectAnyArgsAndReturn( MQTTSuccess );
+    /* PUBLISH failing while processing a second PUBLISH command. */
```
Comment (aggarg) [inline test/unit-test/mqtt_agent_utest.c:1518]:
```suggestion
    /* The stubReceive function ensures that the CommandLoop keeps getting the
     * same command again and again. Ensure that the second PUBLISH fails. */
```

File: source/mqtt_agent.c (hunk 21 above)
Comment (gshvang) [reply source/mqtt_agent.c:660]: Yeah, will address it.

**Review by aggarg — APPROVED** [review]
